### PR TITLE
Add broadcasts M2M on FlowStep, move FlowStart statuses inside class,…

### PR DESCRIPTION
**Pull request by nicpottier**
… move FlowStep types inside class

### CODE REVIEW

File: temba/flows/models.py [hunk 1]
```diff
@@ -2334,190 +2328,649 @@ class Meta:
         ordering = ('-modified_on',)
 
 
-class RuleSet(models.Model):
-
-    TYPE_WAIT_MESSAGE = 'wait_message'
+class FlowRun(models.Model):
```
Comment (nicpottier) [inline temba/flows/models.py]: Had to do some ordering class changes in the file move things inside the classes.

**Comment by nicpottier** [comment]
plz review.. @rowanseymour this has all sorts of stuff you'll love.

**Comment by nicpottier** [comment]
```sql
BEGIN;
CREATE TABLE "flows_flowstep_broadcasts" ("id" serial NOT NULL PRIMARY KEY, "flowstep_id" integer NOT NULL, "broadcast_id" integer NOT NULL, UNIQUE ("flowstep_id", "broadcast_id"));
ALTER TABLE "flows_flowstep_broadcasts" ADD CONSTRAINT "flows_flowstep_flowstep_id_767cf268ab52cf6_fk_flows_flowstep_id" FOREIGN KEY ("flowstep_id") REFERENCES "flows_flowstep" ("id") DEFERRABLE INITIALLY DEFERRED;
ALTER TABLE "flows_flowstep_broadcasts" ADD CONSTRAINT "flows_flowst_broadcast_id_7ec2882a13c82548_fk_msgs_broadcast_id" FOREIGN KEY ("broadcast_id") REFERENCES "msgs_broadcast" ("id") DEFERRABLE INITIALLY DEFERRED;
CREATE INDEX "flows_flowstep_broadcasts_c01a422b" ON "flows_flowstep_broadcasts" ("flowstep_id");
CREATE INDEX "flows_flowstep_broadcasts_b0cb7d59" ON "flows_flowstep_broadcasts" ("broadcast_id");

COMMIT;
```

File: temba/flows/migrations/0055_populate_step_broadcasts.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    for broadcast_batch in chunk_list(batch, 1000):
+        RelatedBroadcast.objects.bulk_create(broadcast_batch)
+
+    # we finished, no need to track any more status
+    r.delete(HIGHPOINT_KEY)
+
+
+def migration_backfill_flowsteps(apps, schema):
+    backfill_flowsteps(apps.get_model('flows', 'FLowStep'),
```
Comment (rowanseymour) [inline temba/flows/migrations/0055_populate_step_broadcasts.py]: Typo

File: temba/flows/migrations/0055_populate_step_broadcasts.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+HIGHPOINT_KEY = 'flowstep_backfill_highpoint'
+
+
+def populate_flowsteps_for_broadcast(RelatedBroadcast, RelatedMsg, MsgManager, broadcast, batch):
+    msg_ids = MsgManager.filter(broadcast=broadcast.id).values_list('id', flat=True)
+    start_count = len(batch)
+
+    for msg_id_batch in chunk_list(set(msg_ids), 1000):
+        fs_ids = set(RelatedMsg.objects.filter(msg_id__in=msg_id_batch).values_list('flowstep_id', flat=True))
```
Comment (rowanseymour) [inline temba/flows/migrations/0055_populate_step_broadcasts.py]: Wouldn't all messages from the same broadcast have the same flowstep ids ?

File: temba/flows/migrations/0055_populate_step_broadcasts.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    broadcast_ids = Broadcast.objects.filter(id__gt=highpoint).order_by('id').values_list('id', flat=True)
+    start = time.time()
+    batch = []
+    i = 0
+
+    for broadcast_id_batch in chunk_list(broadcast_ids, 1000):
+        broadcasts = Broadcast.objects.filter(id__in=broadcast_id_batch).order_by('id').only('id')
+        for broadcast in broadcasts:
+            if i % 1000 == 0:
+                print "Processed %d / %d (batch size %d) in %d" % (i, len(broadcast_ids), len(batch), int(time.time() - start))
```
Comment (rowanseymour) [inline temba/flows/migrations/0055_populate_step_broadcasts.py]: Seems like if this in the outer loop and you don't need the `i % 1000 == 0` stuff, no?

File: temba/flows/migrations/0055_populate_step_broadcasts.py (hunk 3 above)
Comment (nicpottier) [inline temba/flows/migrations/0055_populate_step_broadcasts.py]: Not sure I understand the question. FlowSteps are per contact.. so there's a one to one mapping between a msg and a flowstep. (though potentially more than one msg/broadcast can belong to a flowstep)

File: temba/flows/migrations/0055_populate_step_broadcasts.py (hunk 4 above)
Comment (nicpottier) [inline temba/flows/migrations/0055_populate_step_broadcasts.py]: Oh, so it is.

File: temba/flows/migrations/0055_populate_step_broadcasts.py (hunk 2 above)
Comment (nicpottier) [inline temba/flows/migrations/0055_populate_step_broadcasts.py]: Woah, why does that even work?

File: temba/flows/migrations/0055_populate_step_broadcasts.py (hunk 3 above)
Comment (rowanseymour) [inline temba/flows/migrations/0055_populate_step_broadcasts.py]: In a brief moment of cluelessness I forgot what flowsteps are (not nodes on a flow)

**Pull request merged**